### ISSUE TITLE
OData serialization fails when hiding inherited properties (new modifier)

### ISSUE THREAD

**Issue by howcheng** [OData/WebApi#2403]
When attempting to serialize a derived class that uses the `new` modifier on a property to hide an inherited member, serialization fails due to an AmbiguousMatchException.

### Assemblies affected
Microsoft.AspNet.Odata 7.2.1

### Reproduce steps
Create a class that inherits from another, hiding a property
````
public class Foo
{
    public int Id { get; set; }
}
public class Bar : Foo
{
    public new string Id { get; set; }
}
````
Then create controllers, EDM types, etc, to be able to get a `Bar` object using OData.

### Expected result
Correct serialization
````
{
    "Id" : "123456"
}
`````

### Actual result
`AmbiguousMatchException` is thrown. The stack trace is as follows:
````
System.Reflection.AmbiguousMatchException: Ambiguous match found.
   at System.RuntimeType.GetPropertyImpl(String name, BindingFlags bindingAttr, Binder binder, Type returnType, Type[] types, ParameterModifier[] modifiers)
   at System.Type.GetProperty(String name, BindingFlags bindingAttr)
   at System.Type.GetProperty(String name)
   at Microsoft.AspNet.OData.Formatter.Serialization.TypedEdmStructuredObject.CreatePropertyGetter(Type type, String propertyName)
   at Microsoft.AspNet.OData.Formatter.Serialization.TypedEdmStructuredObject.GetOrCreatePropertyGetter(Type type, String propertyName, IEdmStructuredTypeReference edmType, IEdmModel model)
   at Microsoft.AspNet.OData.Formatter.Serialization.TypedEdmStructuredObject.TryGetPropertyValue(String propertyName, Object& value)
   at Microsoft.AspNet.OData.Query.Expressions.SelectExpandWrapper.TryGetPropertyValue(String propertyName, Object& value)
   at Microsoft.AspNet.OData.ResourceContext.GetPropertyValue(String propertyName)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.CreateStructuralProperty(IEdmStructuralProperty structuralProperty, ResourceContext resourceContext)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.CreateStructuralPropertyBag(SelectExpandNode selectExpandNode, ResourceContext resourceContext)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.CreateResource(SelectExpandNode selectExpandNode, ResourceContext resourceContext)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.WriteResourceAsync(Object graph, ODataWriter writer, ODataSerializerContext writeContext, IEdmTypeReference expectedType)
````

### Additional detail
The line causing the problem is https://github.com/OData/WebApi/blob/master/src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs line 1601
````
object propertyValue = resourceContext.GetPropertyValue(structuralProperty.Name);
````

Unfortunately, I cannot think of a workaround because even though the method `CreateStructuralProperty` is marked virtual, we cannot use `writeContext.GetEdmType` or `serializer.CreateProperty` because both are inaccessible due to protection level.

https://stackoverflow.com/questions/994698/ambiguousmatchexception-type-getproperty-c-sharp-reflection contains a solution for this problem.

**Comment by marabooy** [comment]
Hi @howcheng, the current example above is not supported within OData as this would create some issues within the model creation due to the two types (int & string) being different and not sharing a common parent type. Supporting this could bring about multiple unwanted issues.

**Comment by KenitoInc** [comment]
Fixed by #2523
